File: packages/development/src/specParser.ts
export interface UnknownConfiguration {
    [configPropertyName: string]: string | number | boolean;
}

/** A parsed spec. */
export interface Spec {
    filePath: string;
    message: string;
    fileText: string;
    expectedText: string;
    isOnly: boolean;
    showTree: boolean;
    skip: boolean;
    config: UnknownConfiguration;
}

// todo: write a better parser instead of this lazy stuff

export interface ParseSpecsOptions {
    /** The default file name for a parsed spec. */
    defaultFileName: string;
}

/** Parses all the specs in the given text file. */
export function parseSpecs(fileText: string, options: ParseSpecsOptions) {
    fileText = fileText.replace(/\r?\n/g, "\n");
    const filePath = parseFilePath();
    const configResult = parseConfig();
    const lines = configResult.fileText.split("\n");
    const specStarts = getSpecStarts();
    const specs: Spec[] = [];
    let filterOnly = false;

    for (let i = 0; i < specStarts.length; i++) {
        const startIndex = specStarts[i];
        const endIndex = specStarts[i + 1] || lines.length;
        const messageLine = lines[startIndex];
        const spec = parseSingleSpec(filePath, messageLine, lines.slice(startIndex + 1, endIndex), configResult.config);

        if (spec.skip)
            continue;

        if (spec.isOnly) {
            console.log(`NOTICE!!! Running only test: ${spec.message}`);
            filterOnly = true;
        }

        specs.push(spec);
    }

    return filterOnly ? specs.filter(s => s.isOnly) : specs;

    function getSpecStarts() {
        const result: number[] = [];

        if (!lines[0].startsWith("=="))
            throw new Error("All spec files should start with a message. (ex. == Message ==)");

        for (let i = 0; i < lines.length; i++) {
            if (lines[i].startsWith("=="))
                result.push(i);
        }

        return result;
    }

    function parseFilePath() {
        if (!fileText.startsWith("--"))
            return options.defaultFileName;
        const lastIndex = fileText.indexOf("--\n", 2);
        if (lastIndex === -1)
            throw new Error("Could not find last --\\n.");

        const result = fileText.substring(2, lastIndex).trim();

        // I was lazy while writing this and this is not ideal.
        // This is done for parseConfig()
        fileText = fileText.substring(lastIndex + 3);

        return result;
    }

    function parseConfig(): { fileText: string; config: UnknownConfiguration; } {
        if (!fileText.startsWith("~~"))
            return { fileText, config: {} };
        const lastIndex = fileText.indexOf("~~\n", 2);
        if (lastIndex === -1)
            throw new Error("Canot find last ~~\\n.");
        const configText = fileText.substring(2, lastIndex).replace(/\n/g, "");
        const config: UnknownConfiguration = {};

        for (const item of configText.split(",")) {
            const firstColon = item.indexOf(":");
            const key = item.substring(0, firstColon).trim();
            const value = parseValue(item.substring(firstColon + 1).trim());

            config[key] = value;
        }

        return {
            fileText: fileText.substring(lastIndex + 3),
            config,
        };

        function parseValue(value: string) {
            const parsedInt = parseInt(value, 10);
            if (!isNaN(parsedInt))
                return parsedInt;
            else if (value.startsWith(`"`))
                return value.slice(1, -1); // strip quotes
            else if (value === "true")
                return true;
            else if (value === "false")
                return false;
            return value;
        }
    }
}

function parseSingleSpec(filePath: string, messageLine: string, lines: string[], config: UnknownConfiguration): Spec {
    // this is temp code changed during a port... this should be better
    const fileText = lines.join("\n");
    const parts = fileText.split("[expect]");
    const startText = parts[0].substring(0, parts[0].length - 1); // remove last newline
    const expectedText = parts[1].substring(1, parts[1].length); // remove first newline
    const lowerCaseMessageLine = messageLine.toLowerCase();

    return {
        filePath,
        message: parseMessage(),
        fileText: startText,
        expectedText,
        isOnly: lowerCaseMessageLine.includes("(only)"),
        skip: lowerCaseMessageLine.includes("(skip)"),
        showTree: lowerCaseMessageLine.includes("(tree)"),
        config,
    };

    function parseMessage() {
        // this is very crude...
        return messageLine.substring(2, messageLine.length - 2).trim();
    }
}

File: crates/development/src/spec_parser.rs
use std::collections::HashMap;

#[derive(PartialEq, Debug)]
pub struct Spec {
    pub file_name: String,
    pub message: String,
    pub file_text: String,
    pub expected_text: String,
    pub is_only: bool,
    pub show_tree: bool,
    pub skip: bool,
    pub skip_format_twice: bool,
    pub config: HashMap<String, String>,
}

pub struct ParseSpecOptions {
    /// The default file name for a parsed spec.
    pub default_file_name: &'static str,
}

pub fn parse_specs(file_text: String, options: &ParseSpecOptions) -> Vec<Spec> {
    // this function needs a rewrite
    let file_text = file_text.replace("\r\n", "\n");
    let (file_path, file_text) = parse_file_path(file_text, options);
    let (config, file_text) = parse_config(file_text);
    let lines = file_text.split("\n").collect();
    let spec_starts = get_spec_starts(&file_path, &lines);
    let mut specs = Vec::new();

    for i in 0..spec_starts.len() {
        let start_index = spec_starts[i];
        let end_index = if spec_starts.len() == i + 1 { lines.len() } else { spec_starts[i + 1] };
        let message_line = lines[start_index];
        let spec = parse_single_spec(&file_path, &message_line, &lines[(start_index + 1)..end_index], &config);

        specs.push(spec);
    }

    return specs;

    fn parse_file_path(file_text: String, options: &ParseSpecOptions) -> (String, String) {
        if !file_text.starts_with("--") {
            return (options.default_file_name.into(), file_text);
        }
        let last_index = file_text.find("--\n").expect("Could not find final --");

        (file_text["--".len()..last_index].trim().into(), file_text[(last_index + "--\n".len())..].into())
    }

    fn parse_config(file_text: String) -> (HashMap<String, String>, String) {
        if !file_text.starts_with("~~") {
            return (HashMap::new(), file_text);
        }
        let last_index = file_text.find("~~\n").expect("Could not find final ~~\\n");

        let config_text = file_text["~~".len()..last_index].replace("\n", "");
        let mut config: HashMap<String, String> = HashMap::new();

        for item in config_text.split(",") {
            let first_colon = item.find(":").expect("Could not find colon in config option.");
            let key = item[0..first_colon].trim();
            let value = item[first_colon + ":".len()..].trim();

            config.insert(key.into(), value.into());
        }

        (config, file_text[(last_index + "~~\n".len())..].into())
    }

    fn get_spec_starts(file_name: &str, lines: &Vec<&str>) -> Vec<usize> {
        let mut result = Vec::new();
        let message_separator = get_message_separator(file_name);

        if !lines.first().unwrap().starts_with(message_separator) {
            panic!("All spec files should start with a message. (ex. {0} Message {0})", message_separator);
        }

        for i in 0..lines.len() {
            if lines[i].starts_with(message_separator) {
                result.push(i);
            }
        }

        result
    }

    fn parse_single_spec(file_name: &str, message_line: &str, lines: &[&str], config: &HashMap<String, String>) -> Spec {
        let file_text = lines.join("\n");
        let parts = file_text.split("[expect]").collect::<Vec<&str>>();
        let start_text = parts[0][0..parts[0].len() - "\n".len()].into(); // remove last newline
        let expected_text = parts[1]["\n".len()..].into(); // remove first newline
        let lower_case_message_line = message_line.to_ascii_lowercase();
        let message_separator = get_message_separator(file_name);

        Spec {
            file_name: String::from(file_name),
            message: message_line[message_separator.len()..message_line.len() - message_separator.len()].trim().into(),
            file_text: start_text,
            expected_text,
            is_only: lower_case_message_line.find("(only)").is_some(),
            skip: lower_case_message_line.find("(skip)").is_some(),
            skip_format_twice: lower_case_message_line.find("(skip-format-twice)").is_some(),
            show_tree: lower_case_message_line.find("(tree)").is_some(),
            config: config.clone(),
        }
    }

    fn get_message_separator(file_name: &str) -> &'static str {
        return if file_name.ends_with(".md") {
            "!!"
        } else {
            "=="
        };
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn it_parses() {
        let specs = parse_specs(vec![
            "== message 1 ==",
            "start",
            "multiple",
            "",
            "[expect]",
            "expected",
            "multiple",
            "",
            "== message 2 (only) (tree) (skip) (skip-format-twice) ==",
            "start2",
            "",
            "[expect]",
            "expected2",
            "",
        ].join("\n"), &ParseSpecOptions { default_file_name: "test.ts" });

        assert_eq!(specs.len(), 2);
        assert_eq!(specs[0], Spec {
            file_name: "test.ts".into(),
            file_text: "start\nmultiple\n".into(),
            expected_text: "expected\nmultiple\n".into(),
            message: "message 1".into(),
            is_only: false,
            show_tree: false,
            skip: false,
            skip_format_twice: false,
            config: HashMap::new(),
        });
        assert_eq!(specs[1], Spec {
            file_name: "test.ts".into(),
            file_text: "start2\n".into(),
            expected_text: "expected2\n".into(),
            message: "message 2 (only) (tree) (skip) (skip-format-twice)".into(),
            is_only: true,
            show_tree: true,
            skip: true,
            skip_format_twice: true,
            config: HashMap::new(),
        });
    }

    #[test]
    fn it_parses_with_file_name() {
        let specs = parse_specs(vec![
            "-- asdf.ts --",
            "== message ==",
            "start",
            "[expect]",
            "expected",
        ].join("\n"), &ParseSpecOptions { default_file_name: "test.ts" });

        assert_eq!(specs.len(), 1);
        assert_eq!(specs[0], Spec {
            file_name: "asdf.ts".into(),
            file_text: "start".into(),
            expected_text: "expected".into(),
            message: "message".into(),
            is_only: false,
            show_tree: false,
            skip: false,
            skip_format_twice: false,
            config: HashMap::new(),
        });
    }

    #[test]
    fn it_parses_with_config() {
        let specs = parse_specs(vec![
            "-- asdf.ts --",
            "~~ test.test: other, lineWidth: 40 ~~",
            "== message ==",
            "start",
            "[expect]",
            "expected",
        ].join("\n"), &ParseSpecOptions { default_file_name: "test.ts" });

        assert_eq!(specs.len(), 1);
        assert_eq!(specs[0], Spec {
            file_name: "asdf.ts".into(),
            file_text: "start".into(),
            expected_text: "expected".into(),
            message: "message".into(),
            is_only: false,
            show_tree: false,
            skip: false,
            skip_format_twice: false,
            config: [("test.test".into(), "other".into()), ("lineWidth".into(), "40".into())].iter().cloned().collect(),
        });
    }

    #[test]
    fn it_parses_markdown() {
        let specs = parse_specs(vec![
            "!! message 1 !!",
            "start",
            "multiple",
            "",
            "[expect]",
            "expected",
            "multiple",
            "",
            "!! message 2 (only) (tree) (skip) (skip-format-twice) !!",
            "start2",
            "",
            "[expect]",
            "expected2",
            "",
        ].join("\n"), &ParseSpecOptions { default_file_name: "test.md" });

        assert_eq!(specs.len(), 2);
        assert_eq!(specs[0], Spec {
            file_name: "test.md".into(),
            file_text: "start\nmultiple\n".into(),
            expected_text: "expected\nmultiple\n".into(),
            message: "message 1".into(),
            is_only: false,
            show_tree: false,
            skip: false,
            skip_format_twice: false,
            config: HashMap::new(),
        });
        assert_eq!(specs[1], Spec {
            file_name: "test.md".into(),
            file_text: "start2\n".into(),
            expected_text: "expected2\n".into(),
            message: "message 2 (only) (tree) (skip) (skip-format-twice)".into(),
            is_only: true,
            show_tree: true,
            skip: true,
            skip_format_twice: true,
            config: HashMap::new(),
        });
    }
}
